### ISSUE TITLE
dart2js: remove methods that delegate to super-method

### ISSUE THREAD

**Issue by rakudrama** [dart-lang/sdk#35917]
If all a method does is delegate to its super-method, the method can be removed, relying on the property-based dispatch to execute the super-method directly. The Flutter framework `toString()` methods have this property. Let us call these 'trivial' methods.

The requirement is that the optimization applies to methods containing assertions and other debug code that gets optimized away, leaving only a supercall. We can't tell prior to optimization that the method is trivial (and this is the whole point, otherwise the method would not have been written!)

Recognizing a trivial method is not straight-forward. The generated JavaScript does not have sufficient information to distinguish a trivial method from other methods. We would have to recognize trivial methods with a CFG pass just prior to JavaScript codegen. The cost of an extra pass can be ameliorated by detecting a super-call in an earlier pass.

Making the program work after the trivial method is removed is tricky because function bodies are compiled independently (for scalability). The problem is that subclasses may target the method we want to remove, but the super calls in the subclasses can't tell until too late that the target function has disappeared. It is impossible to go back and 'fix' the method that references one we removed.

```dart
class A {
  toString() => debugMode ? 'A' : super.toString();
}
class B extends A {
}
class C extends B {
  toString() => 'C:' + super.toString();
}
```

Typically `C.toString` will call a member via direct lookup
```js
C.prototype = {
  toString$0: function() {
    return 'C:' + A.prototype.toString$0.call(this);
  }
}
```
or via an alias
```js
    return 'C:' + this.super$A$toString$0()
```
where alias is created via an assignment prior to the prototypes being linked for inheritance
```js
    A.prototype.super$A$toString$0 = A.prototype.toString$0
```
(dart2js currently uses both mechanisms, although the alias one has better performance.)

An additional source of references to trivial methods is tear-offs. Consider the expression `super.toString` in class `B` or `C` - that would be a call to the super-getter that creates a tear-off for the super-method.

I think the alias mechanism could be made to work for removing more trivial methods since the list of aliases is generated later, on the basis of calls using the aliases.

We can remove a trivial method that _has no references_. This would only be leaf trivial methods because trivial methods themselves contain references. We could account for this by tracking references from within trivial methods separately to other methods, and removing the references due to the removed trivial methods.  This approach would 'unpeel' the super-chain from the leaf class until a non-trivial method, but not remove any trivial calls that have non-trivial overrides.
To remove other trivial methods we would have to remove the references redirecting them to the non-trivial method, i.e.
```js
    A.prototype.super$A$toString$0 = A.prototype.toString$0
-->
    Object.prototype.super$A$toString$0 = Object.prototype.toString$0
```

JavaScript ES6 has a `super`-call feature which we don't use because it is not supported by IE11. super-calls also require the concept of a [[HomeObject]], which limits the JavaScript syntax available for creating instance methods. What is nice about ES6 `super` is that it makes the call-chain references implicit. We can simply remove trivial methods that have no tear-off getter.

**Comment by rakudrama** [comment]
TL;DR: Given the complexity, special case nature, and relatively low impact of the optimization I think we should defer working on it until we are making better use of ES6.

/cc @ferhatb @sigmundch @vsmenon

**Comment by sigmundch** [comment]
Thanks for the detailed analysis - for the specific case of `toString`, I wonder then if we should instead push for the annotation-based tree-shaking approach (already tracked by https://github.com/dart-lang/sdk/issues/35303).

**Comment by rakudrama** [comment]
I looked at the flutter gallery and found there are 68 `toString` methods that directly call the super method after optimization. If all of these methods could simply be deleted, the code size saving would be ~0.12%.
